### PR TITLE
Change Function registry flow; Get rid of whole-archive in compile

### PR DESCRIPTION
**Pull request by raymondxyang**
* Change Function registry flow into the Opschema registering way.
* Get rid of whole-archive during compiling

### CODE REVIEW

File: onnx/defs/experiments/experiments_functions.cc [hunk 1]
```diff
@@ -171,5 +171,11 @@ static Common::Status BuildMVN(std::unique_ptr<FunctionProto>* func_proto) {
...
 }
 
-ONNX_FUNCTION(
+ONNX_FUNCTION_BUILD(
+    MeanVarianceNormalization,
+    9,
     FunctionBuilder().SetDomain(ONNX_DOMAIN).SetBuildFunction(BuildMVN));
+
+} // namespace ONNX_NAMESPACE
+// ONNX_FUNCTION(
```
Comment (houseroad) [inline onnx/defs/experiments/experiments_functions.cc]: clean up?

Comment (raymondxyang) [reply onnx/defs/experiments/experiments_functions.cc]: This means to leave a example for registering functions. I ll update the marco to be the latest one.

**Comment by raymondxyang** [comment]
@RyanUnderhill please help take a look thanks😄

File: onnx/defs/function.cc [hunk 2]
```diff
@@ -46,6 +47,17 @@ Common::Status FunctionBuilderRegistry::GetFunctions(
...
 
+  class FunctionBuilderRegisterer {
+   public:
+    FunctionBuilderRegisterer() {
+      RegisterOnnxFunctionBuilder();
+    }
+  };
+
+#ifndef __ONNX_DISABLE_STATIC_REGISTRATION
+  static FunctionBuilderRegisterer functionBuilder_registerer;
```
Comment (RyanUnderhill) [inline onnx/defs/function.cc]: If you want to remove the dummy class, you can write this:

`static bool functionBuilder_registerer=(RegisterOnnxFunctionBuilder(), false);`

File: onnx/defs/function.h [hunk 3]
```diff
@@ -62,6 +62,25 @@ class FunctionBuilderRegistry : public IFunctionBuilderRegistry {
...
+
+#define ONNX_FUNCTION_BUILDER_CLASS_NAME(domain, ver, name) \
+  name##_##domain##_ver##ver
+
+#define ONNX_FUNCTION_BUILD(name, ver, build_func) \
+  ONNX_FUNCTION_BUILD_HELPER(                      \
+      __COUNTER__, name, Onnx, ONNX_DOMAIN, ver, build_func)
+
+#define ONNX_FUNCTION_BUILD_HELPER(                                           \
+    counter, name, domain, domain_str, ver, build_func)                       \
```
Comment (jeffbloo) [inline onnx/defs/function.h]: The counter parameter can be removed here

**Pull request merged**